cassiopeia/solver/PercolationSolver.py

```diff
@@ -0,0 +1,350 @@
+                n for n in lca_network if lca_network.in_degree(n) == 0
+            ][0]
+
+            # Take the split at the root as the clusters of components
+            # in the split, ignoring unifurcations
+            current_node = root
+            while len(grouped_components) == 0:
+                successors = list(lca_network.successors(current_node))
+                if len(successors) == 1:
+                    current_node = successors[0]
+                else:
+                    for i in successors:
+                        grouped_components.append(
+                            solver_utilities.get_leaf_children(
+                                lca_network, i
+                            )
+                        )
+
+            # For each component in each cluster, take the nodes in that
+            # component to form the final split
+            for cluster in grouped_components:
+                sample_index_group = []
+                for component in cluster:
+                    sample_index_group.extend(component_to_nodes[component])
+                partition_sides.append(sample_index_group)
+        else:
+            for c in range(len(connected_components)):
+                partition_sides.append(list(connected_components[c]))
+
+        # Convert from component indices back to the sample names in the 
+        # original character matrix
+        sample_names = list(character_matrix.index)
+        partition_named = []
+        for sample_index_group in partition_sides:
+            sample_name_group = []
+            for sample_index in sample_index_group:
+                sample_name_group.append(
+                    sample_names[sample_index]
+                )
+            partition_named.append(sample_name_group)
+
+        return partition_named
+
+
+    def __add_duplicates_to_tree(
+        self,
+        tree: nx.DiGraph,
+        character_matrix: pd.DataFrame
+    ) -> nx.DiGraph:
+        """Takes duplicate samples and places them in the tree.
+
+        Places samples removed in removing duplicates in the tree as sisters
+        to the corresponding cells that share the same mutations.
+
+        Args:
+            tree: The tree to have duplicates added to
+            character_matrix: Character matrix
+
+        Returns:
+            A tree with duplicates added
+        """
+
+        duplicate_groups = (
+            character_matrix[character_matrix.duplicated(keep=False) == True]
+            .reset_index()
+            .groupby(character_matrix.columns.tolist())["index"]
+            .agg(["first", tuple])
+            .set_index("first")["tuple"]
+            .to_dict()
+        )
+
+        for i in duplicate_groups:
+            if len(tree.nodes) == 1:
+                new_internal_node = len(duplicate_groups[i]) + 1
+            else:
+                new_internal_node = (
+                    max([n for n in tree.nodes if type(n) == int]) + 1
+                )
```

Comment:
Same comment as above- why not use `new_internal_node = len(tree.nodes)+1`?

Reply:
See above comment for the GreedySolver